components/eamxx/src/share/util/eamxx_time_interpolation.cpp

```diff
@@ -0,0 +1,343 @@
+/* Function to organize the data available from data files by time.  This is necessary to quickly
+ * grab the appropriate data each time interpolation is requested.
+ * Input:
+ *   list_of_files - Is a vector of strings representing all files that can be used for data.
+ *
+ * We create a reference timestamp to use when sorting the data snaps.
+ */
+void TimeInterpolation::set_file_data_triplets(const vos_type& list_of_files) {
+  EKAT_REQUIRE_MSG(list_of_files.size()>0,"ERROR! TimeInterpolation::set_file_data_triplets - the list of files is empty. Please check.");
+  TimeStamp ts_ref;
+  // The first step is to grab all relevant metadata for the DataFromFileTriplet objects.
+  // We will store the times in a map and take advantage of maps natural sorting to organize the triplets
+  // in chronological order.  This ensures that if the list of files does not represent the chronological
+  // order to the data we will still have sorted data.
+  vos_type               filenames_tmp;
+  std::vector<TimeStamp> timestamps_tmp;
+  std::vector<int>       time_idx_tmp;
+  std::map<int,int>      map_of_times_to_vector_idx;
+  int running_idx = 0;
+  for (int ii=0; ii<list_of_files.size(); ii++) {
+    const auto filename = list_of_files[ii];
+    // Reference TimeStamp
+    const int date_start = scorpio::get_attribute<int>(filename,"start_date"); // Start date is in YYYYMMDD format
+    const int time_start = scorpio::get_attribute<int>(filename,"start_time"); // Start time is in hhmmss format
+    // Need to parse the start time and date into a timestamp
+    const int YY = date_start/10000;
+    const int MM = (date_start - YY*10000)/100;
+    const int DD = (date_start - YY*10000 - MM*100);
+    const int hh = time_start/10000;
+    const int mm = (time_start - hh*10000)/100;
+    const int ss = (time_start - hh*10000 - mm*100);
+    TimeStamp ts_file_start(YY,MM,DD,hh,mm,ss);
+    if (ii==0) {
+      ts_ref = ts_file_start;
+    }
+    // Gather information about time in this file
+    scorpio::register_file(filename,scorpio::Read);
+    const int ntime = scorpio::get_dimlen(filename,"time");
+    for (int tt=0; tt<ntime; tt++) {
+      auto time_snap = scorpio::read_time_at_index_c2f(filename.c_str(),tt+1);
+      TimeStamp ts_snap = ts_file_start;
+      if (time_snap>0) {
+        ts_snap += (time_snap*86400); // note, time is assumed to be in days.
+      }
+      auto time = ts_snap.seconds_from(ts_ref);
+      map_of_times_to_vector_idx.emplace(time,running_idx);
+      filenames_tmp.push_back(filename);
+      timestamps_tmp.push_back(ts_snap);
+      time_idx_tmp.push_back(tt);
+      ++running_idx;
+    }
+  }
+  // Now that we have gathered all of the timesnaps we can arrange them in order as DataFromFileTriplet objects.
+  // Taking advantage of maps automatically self-sorting by the first arg.
+  for (auto a : map_of_times_to_vector_idx) {
```

Comment:
I think all you need is to store the first (and last, for safety) timestamp in each file. Then, using std::upper_bound or std::lower_bound, you can efficiently find which file you need to look in.

Reply:
Also, if we can assume we will alwasy "move forward" with time, and that our timestep will not be so large as to skip a file, then we don't really need to search.

Reply:
I didn't want to assume that our timestep will always be smaller than the data.  I think 99% of the time that will be true.  But there is no guarantee that this is the case.

Reply:
i don't see how we avoid a search one way or the other.  If we store upper and lower bounds that will make it easy to quickly know which file grab, but then we will still need to search through the `time` vector of that file to figure out which snap to take.

Unless storing the triplet information is expensive, in which case it would probably be better to search on demand rather then preorganize all the triplets.

Reply:
The way it is now, we store a triplet for each timestamp, which seems an overkill. Storing t0,tN for each file, and the timestamps in each file should be enough. Using something like std::upper_bound (twice), we should be able to a) locate the file, and b) locate the time index.

If you want, I can take a stab at this (next week).